File: java/BillsApi/src/main/java/com/apilangarena/javabillsapi/infrastructure/persistence/BillWriteRepositoryImpl.java
package com.apilangarena.javabillsapi.infrastructure.persistence;

import com.apilangarena.javabillsapi.application.bills.create.BillWriteRepository;
import com.apilangarena.javabillsapi.application.bills.create.CreateBillResult;
import com.apilangarena.javabillsapi.domain.bills.NewBill;
import com.apilangarena.javabillsapi.domain.bills.NewBillLine;
import com.apilangarena.javabillsapi.infrastructure.persistence.entity.BillEntity;
import com.apilangarena.javabillsapi.infrastructure.persistence.entity.BillLineEntity;
import com.apilangarena.javabillsapi.infrastructure.persistence.repository.BillJpaRepository;
import org.springframework.stereotype.Repository;

import java.util.Objects;

@Repository
public class BillWriteRepositoryImpl implements BillWriteRepository {
    private final BillJpaRepository billJpaRepository;

    public BillWriteRepositoryImpl(BillJpaRepository billJpaRepository) {
        this.billJpaRepository = billJpaRepository;
    }

    @Override
    public boolean existsByBillNumber(String billNumber) {
        return billJpaRepository.existsByBillNumber(billNumber);
    }

    @Override
    public CreateBillResult create(NewBill newBill) {
        BillEntity entity = new BillEntity(
            newBill.getBillNumber(),
            newBill.getIssuedAt(),
            newBill.getCustomerName(),
            newBill.getSubtotal(),
            newBill.getTax(),
            newBill.getCurrency()
        );

        for (NewBillLine line : newBill.getLines()) {
            entity.addLine(new BillLineEntity(
                line.getLineNo(),
                line.getConcept(),
                line.getQuantity(),
                line.getUnitAmount(),
                line.getLineAmount()
            ));
        }

        BillEntity saved = billJpaRepository.saveAndFlush(entity);
        Long savedId = Objects.requireNonNull(saved.getId(), "Persisted bill id was null.");

        return new CreateBillResult(
            savedId,
            saved.getBillNumber(),
            saved.getIssuedAt(),
            newBill.getSubtotal(),
            newBill.getTax(),
            newBill.getTotal(),
            saved.getCurrency()
        );
    }
}

File: kotlin/BillsApi/src/main/kotlin/com/apilangarena/billsapi/infrastructure/persistence/BillWriteRepositoryImpl.kt
package com.apilangarena.billsapi.infrastructure.persistence

import com.apilangarena.billsapi.application.bills.create.BillWriteRepository
import com.apilangarena.billsapi.application.bills.create.CreateBillResult
import com.apilangarena.billsapi.domain.bills.NewBill
import com.apilangarena.billsapi.infrastructure.persistence.entity.BillEntity
import com.apilangarena.billsapi.infrastructure.persistence.entity.BillLineEntity
import com.apilangarena.billsapi.infrastructure.persistence.repository.BillJpaRepository
import org.springframework.stereotype.Repository

@Repository
class BillWriteRepositoryImpl(
    private val billJpaRepository: BillJpaRepository,
) : BillWriteRepository {
    override fun existsByBillNumber(billNumber: String): Boolean {
        return billJpaRepository.existsByBillNumber(billNumber)
    }

    override fun create(newBill: NewBill): CreateBillResult {
        val entity = BillEntity(
            billNumber = newBill.billNumber,
            issuedAt = newBill.issuedAt,
            customerName = newBill.customerName,
            subtotal = newBill.subtotal,
            tax = newBill.tax,
            currency = newBill.currency,
        )

        newBill.lines.forEach { line ->
            entity.addLine(
                BillLineEntity(
                    lineNo = line.lineNo,
                    concept = line.concept,
                    quantity = line.quantity,
                    unitAmount = line.unitAmount,
                    lineAmount = line.lineAmount,
                )
            )
        }

        val saved = billJpaRepository.saveAndFlush(entity)
        val savedId = requireNotNull(saved.id) { "Persisted bill id was null." }

        return CreateBillResult(
            id = savedId,
            billNumber = saved.billNumber,
            issuedAt = saved.issuedAt,
            subtotal = newBill.subtotal,
            tax = newBill.tax,
            total = newBill.total,
            currency = saved.currency,
        )
    }
}
